server/src/main/java/com/flipkart/varadhi/web/v1/authz/IamPolicyHandlers.java

```diff
@@ -39,58 +43,60 @@ public IamPolicyHandlers(ProjectService projectService, IamPolicyService iamPoli
+                    ctx.pathParam(PATH_PARAM_SUBSCRIPTION)
+            );
+            case IAM_POLICY -> throw new IllegalArgumentException("Iam Policy is not a resource");
+
+        };
     }
 
-    public ResourceHierarchy getIamHierarchy(RoutingContext ctx, boolean hasBody) {
-        String resourceType = ctx.request().getParam("resource_type");
-        String resourceName = ctx.request().getParam("resource");
-        if (null != resourceType && null != resourceName) {
-            return new Hierarchies.IamPolicyHierarchy(resourceType, resourceName);
-        }
-        return new Hierarchies.RootHierarchy();
+    private static ResourceAction getActionBasedOnResource(ResourceType type) {
+        return switch (type) {
+            case TOPIC -> TOPIC_IAM_POLICY_GET;
+            case SUBSCRIPTION -> SUBSCRIPTION_IAM_POLICY_GET;
+            default -> IAM_POLICY_GET;
+        };
     }
 
-    private List<RouteDefinition> getPolicyHandlers() {
-        return new SubRoutes(
-                "/v1",
-                List.of(
-                        RouteDefinition.get("GetIAMPolicy", "/orgs/:org/policy")
-                                .build(this::getHierarchy, this.getIAMPolicyHandler(ResourceType.ORG)),
-                        RouteDefinition.put("SetIAMPolicy", "/orgs/:org/policy")
-                                .hasBody()
-                                .build(this::getHierarchy, this.setIAMPolicyHandler(ResourceType.ORG)),
-                        RouteDefinition.get("GetIAMPolicy", "/orgs/:org/teams/:team/policy")
-                                .build(this::getHierarchy, this.getIAMPolicyHandler(ResourceType.TEAM)),
-                        RouteDefinition.put("SetIAMPolicy", "/orgs/:org/teams/:team/policy")
-                                .hasBody()
-                                .build(this::getHierarchy, this.setIAMPolicyHandler(ResourceType.TEAM)),
-                        RouteDefinition.get("GetIAMPolicy", "/projects/:project/policy")
-                                .build(this::getHierarchy, this.getIAMPolicyHandler(ResourceType.PROJECT)),
-                        RouteDefinition.put("SetIAMPolicy", "/projects/:project/policy")
-                                .hasBody()
-                                .build(this::getHierarchy, this.setIAMPolicyHandler(ResourceType.PROJECT)),
-                        RouteDefinition.get("GetIAMPolicy", "/projects/:project/topics/:topic/policy")
-                                .build(this::getHierarchy, this.getIAMPolicyHandler(ResourceType.TOPIC)),
-                        RouteDefinition.put("SetIAMPolicy", "/projects/:project/topics/:topic/policy")
-                                .hasBody()
-                                .build(this::getHierarchy, this.setIAMPolicyHandler(ResourceType.TOPIC))
-                )
-        ).get();
+    private static ResourceAction setActionBasedOnResource(ResourceType type) {
```

Comment:
every time having to refactor this is too tedious with the route definition contracts frequently evolving with more usage.
this will allow for minimal changes keeping the effort low